modules/client/src/main/java/org/apache/ignite/internal/client/table/ClientTableMapUtils.java

```diff
@@ -0,0 +1,242 @@
+                resFut.complete(in);
+            } else {
+                R finalIn = in;
+                CompletableFutures.allOf(waitCommitFuts).whenComplete((r, e) -> {
+                    // Ignore errors.
+                    resFut.complete(finalIn);
+                });
+            }
+
+            return null;
+        });
+    }
+
+    static <E> void mapAndRetry(
+            MapFunction<E, List<E>> mapFun,
+            Collection<E> keys,
+            List<Transaction> txns,
+            Map<Integer, Batch<E>> mapped,
+            long[] startTs,
+            CompletableFuture<List<E>> resFut,
+            IgniteLogger log
+    ) {
+        if (startTs[0] == 0) {
+            startTs[0] = System.nanoTime();
+        }
+
+        List<CompletableFuture<List<E>>> res = new ArrayList<>(mapped.size());
+        List<Batch<E>> batches = new ArrayList<>(mapped.size());
+
+        for (Entry<Integer, Batch<E>> entry : mapped.entrySet()) {
+            res.add(mapFun.apply(entry.getValue().batch, PartitionAwarenessProvider.of(entry.getKey()), mapped.size() > 1));
+            batches.add(entry.getValue());
+        }
+
+        CompletableFutures.allOf(res).handle((ignored, err) -> {
+            List<CompletableFuture<Void>> waitCommitFuts = List.of();
+            if (!txns.isEmpty()) {
+                if (err != null) {
+                    boolean needRetry = rollbackIfNeeded(txns, res, log);
+
+                    long nowRelative = System.nanoTime();
+                    if (needRetry && nowRelative - startTs[0] < DEFAULT_IMPLICIT_GET_ALL_TIMEOUT_NANOS) {
+                        startTs[0] = nowRelative;
+                        txns.clear(); // The collection is re-filled on next map attempt.
+
+                        mapAndRetry(mapFun, keys, txns, mapped, startTs, resFut, log);
+
+                        return null;
+                    }
+
+                    resFut.completeExceptionally(err);
+
+                    return null;
+                }
+
+                waitCommitFuts = unlockFragments(txns, log);
+            } else {
+                if (err != null) {
+                    resFut.completeExceptionally(err);
+
+                    return null;
+                }
+            }
+
+            var in = new ArrayList<E>(Collections.nCopies(keys.size(), null));
+
+            for (int i = 0; i < res.size(); i++) {
+                CompletableFuture<List<E>> f = res.get(i);
+                reduceWithKeepOrder(in, f.getNow(null), batches.get(i).originalIndices);
+            }
+
+            if (waitCommitFuts.isEmpty()) {
+                resFut.complete(in);
+            } else {
+                CompletableFutures.allOf(waitCommitFuts).whenComplete((r, e) -> {
+                    // Ignore errors.
+                    resFut.complete(in);
+                });
+            }
+
+            return null;
+        });
+    }
+
+    private static <E> boolean rollbackIfNeeded(
```

Comment:
Why does the method call `rollbackIfNeeded` because it rolls back all transactions in any case?